jax/_src/lax/parallel.py

```diff
@@ -1094,6 +1177,94 @@ def _raise_to_shaped_abstract_eval(x, *, axis_name, **params):
 batching.fancy_primitive_batchers[ppermute_p] = _ppermute_batcher
 batching.skippable_batchers[ppermute_p] = partial(_names_in_param, 'axis_name')
 
+
+def _psend_lowering(ctx, x, *, axis_name, perm):
+  if ("cuda" not in ctx.module_context.platforms):
+    raise NotImplementedError("psend is currently only implemented on GPUs")
+
+  full_perm, other_args = _pcollectives_lowering_common(
+      ctx, axis_name=axis_name, perm=perm, op_name="psend"
+  )
+  token = hlo.create_token()
+  send_op = hlo.SendOp(
+      [x],
+      token,
+      source_target_pairs=mlir.dense_int_elements(full_perm),
+      **other_args,
+  )
+  axis_ctx = ctx.module_context.axis_context
+  if isinstance(axis_ctx, ShardingContext):
+    raise NotImplementedError("psend currently only supports manual sharding")
+  assert isinstance(axis_ctx, SPMDAxisContext)
+
+  sharding = xc.OpSharding()
+  sharding.type = xc.OpSharding.Type.MANUAL
+  mlir.set_sharding(send_op, sharding)
+  return [send_op.results]
+
+
+mlir.lowerable_effects.add_type(core.SingleSideCollectiveEffect)
+
+
+def _psend_abstract_eval(x, *, axis_name, **params):
+  _check_axis_names(axis_name)
+  return abstract_token, {
+      *map(core.NamedAxisEffect, axis_name),
+      core.SingleSideCollectiveEffect(),
+  }
+
+
+psend_p = core.Primitive("psend")
+psend_p.def_impl(partial(dispatch.apply_primitive, psend_p))
+psend_p.def_effectful_abstract_eval(_psend_abstract_eval)
+mlir.register_lowering(psend_p, _psend_lowering)
+batching.fancy_primitive_batchers[psend_p] = _ppermute_batcher
+batching.skippable_batchers[psend_p] = partial(_names_in_param, "axis_name")
+
+
+def _precv_lowering(ctx, token, *, out_shape, axis_name, perm):
+  if ("cuda" not in ctx.module_context.platforms):
```

Comment:
There's a better way to write this:

```
def _precv_lowering_gpu(...):
  ...

mlir.register_lowering(precv_p, _precv_lowering_gpu, platform='gpu')

def _precv_lowering(...):
  raise NotImplementedError(...)
mlir.register_lowering(precv_p, _precv_lowering)
```

Same for psend_p

Reply:
suggestion applied